packages/dev/core/src/AudioV2/abstractAudio/abstractAudioNode.ts

```diff
@@ -0,0 +1,195 @@
+
+        if (nodeType | _AudioNodeType.In) {
+            this._downstreamNodes = new Set<AbstractAudioNode>();
+        }
+
+        if (nodeType | _AudioNodeType.Out) {
+            this._upstreamNodes = new Set<AbstractAudioNode>();
+        }
+    }
+
+    /**
+     * Releases associated resources.
+     * - Triggers `onDisposeObservable`.
+     * @see {@link onDisposeObservable}
+     */
+    public dispose(): void {
+        if (this._downstreamNodes) {
+            for (const node of Array.from(this._downstreamNodes)) {
+                this._disconnect(node);
+            }
+            this._downstreamNodes.clear();
+        }
+
+        if (this._upstreamNodes) {
+            for (const node of Array.from(this._upstreamNodes)) {
+                node._disconnect(this);
+            }
+            this._upstreamNodes.clear();
+        }
+
+        this.onDisposeObservable.notifyObservers(this);
+        this.onDisposeObservable.clear();
+    }
+
+    /**
+     * Gets a string identifying the name of the class
+     * @returns the class's name as a string
+     */
+    public abstract getClassName(): string;
+
+    /**
+     * Connect to a downstream audio input node.
+     * @param node - The downstream audio input node to connect
+     */
+    protected _connect(node: AbstractAudioNode): void {
+        if (!this._downstreamNodes) {
+            return;
+        }
+
+        if (this._downstreamNodes.has(node)) {
+            return;
+        }
+
+        if (!node._onConnect(this)) {
+            return;
+        }
+
+        this._downstreamNodes.add(node);
+    }
+
+    /**
+     * Disconnects a downstream audio input node.
+     * @param node - The downstream audio input node to disconnect
+     */
+    protected _disconnect(node: AbstractAudioNode): void {
+        if (!this._downstreamNodes) {
+            return;
+        }
+
+        this._downstreamNodes.delete(node);
+
+        node._onDisconnect(this);
+    }
+
+    /**
+     * Called when an upstream audio output node is connecting.
+     * @param node - The connecting upstream audio node
+     * @returns `true` if the connection succeeds; otherwise `false`
+     */
+    private _onConnect(node: AbstractAudioNode): boolean {
+        if (!this._upstreamNodes) {
+            return false;
+        }
+
+        this._upstreamNodes.add(node);
+
+        return true;
+    }
+
+    /**
+     * Called when an upstream audio output node disconnects.
+     * @param node - The disconnecting upstream audio node
+     */
+    private _onDisconnect(node: AbstractAudioNode): void {
+        this._upstreamNodes?.delete(node);
+    }
+}
+
+/**
+ * Abstract class for a named audio node.
+ */
+export abstract class AbstractNamedAudioNode extends AbstractAudioNode {
+    private _name: string;
+
+    /**
+     * Observable for when the audio node is renamed.
+     */
+    public readonly onNameChangedObservable = new Observable<{ oldName: string; node: AbstractNamedAudioNode }>();
```

Comment:
I assume you don't want to provide "newName" because it i a part of the node?

Reply:
Ya, that's the idea, but maybe `newName` should be a parameter, too? Seems like it should since it caught your eye.

Reply:
I like providing the parameters needed directly, but it is a matter of taste TBH. it can be seen as redundant as you pass the new node.

Reply:
Ya, I've been on the fence for this but I think I prefer the parameter, too. It makes it more obvious and discoverable.